server/resp/src/main/java/org/infinispan/server/resp/scripting/LuaTaskEngine.java

```diff
@@ -0,0 +1,380 @@
+package org.infinispan.server.resp.scripting;
+
+import static org.infinispan.server.resp.scripting.LuaContext.sha1hex;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletionStage;
+
+import org.infinispan.commons.CacheException;
+import org.infinispan.commons.util.concurrent.CompletableFutures;
+import org.infinispan.scripting.ScriptingManager;
+import org.infinispan.scripting.impl.ScriptMetadata;
+import org.infinispan.scripting.impl.ScriptWithMetadata;
+import org.infinispan.server.resp.Resp3Handler;
+import org.infinispan.server.resp.serialization.ResponseWriter;
+import org.infinispan.server.resp.serialization.SerializationHint;
+import org.infinispan.server.resp.serialization.lua.LuaResponseWriter;
+import org.infinispan.server.resp.tx.TransactionContext;
+import org.infinispan.tasks.Task;
+import org.infinispan.tasks.TaskContext;
+import org.infinispan.tasks.spi.TaskEngine;
+import org.infinispan.util.concurrent.BlockingManager;
+
+import io.netty.channel.ChannelHandlerContext;
+import party.iroiro.luajava.Lua;
+import party.iroiro.luajava.lua51.Lua51Consts;
+
+/**
+ * An Infinispan TaskEngine built specifically for executing lua scripts in the context of the resp connector.
+ * It is therefore not a generic task engine or a scripting engine that can be used from Hot Rod or REST.
+ */
+public class LuaTaskEngine implements TaskEngine {
+   private final LuaContextPool pool;
+   private final ScriptingManager scriptingManager;
+
+   public LuaTaskEngine(ScriptingManager scriptingManager) {
+      this.scriptingManager = scriptingManager;
+      this.pool = new LuaContextPool(LuaContext::new, 2, 4);
+   }
+
+   public void shutdown() {
+      pool.shutdown();
+   }
+
+   public CompletionStage<Void> eval(Resp3Handler handler, ChannelHandlerContext ctx, String code, String[] keys, String[] args, long flags) {
+      return handler.getBlockingManager().supplyBlocking(() -> {
+         LuaContext luaCtx = pool.borrow();
+         LuaCode script = scriptLoad(code, false);
+         try {
+            luaCtx.registerScript(script);
+            runScript(luaCtx, handler, ctx, script, keys, args, flags);
+         } finally {
+            luaCtx.unregisterScript(script);
```

Comment:
Do we need to close the luaCtx if there was an exception? It won't be returned to the pool and the other stuff in `thenApplyAsync` won't run either.

Reply:
I've reworked this